### ISSUE TITLE
inaccessible *ebiten.DrawImageOptions

### ISSUE THREAD

**Issue by setanarut** [yohamta/ganim8#22]
*ebiten.DrawImageOptions is the most important struct in ebitengine. It needs to be accessible to work with 2D camera systems.

https://github.com/yohamta/ganim8/blob/5328f97f11666ae6d0a2e570995835c25470462d/sprite.go#L161C15-L161C15

**Comment by setanarut** [comment]
Animation player should not be responsible for the screen drawing. It only needs to return `*ebiten.Image` with changing index.

**Comment by yohamta** [comment]
I completely agree with it, let’s discuss the new API. Maybe we could add it as separate interface and then clean up as v3.

**Comment by yohamta** [comment]
Sorry my finger slipped

**Comment by setanarut** [comment]
It will take time to understand how this package works. I am thinking of writing my own animation player. Based on this example. https://ebitengine.org/en/examples/animation.html
I will try if it can be done without using the `time` package.

```Go
package main

import "fmt"

func main() {
	for gameFrame := 0; gameFrame < 50; gameFrame++ {
		fmt.Print(getFrameIndex(gameFrame, 7, 3))
	}
}

func getFrameIndex(gameFrame, totalFrames, repeat int) int {
	frameIndex := (gameFrame / repeat) % totalFrames
	return frameIndex
}

// output:
// 0011223344556600112233445566001122334455660011223344556600112233445566001122334455660011223344556600
```

**Comment by yohamta** [comment]
I would greatly appreciate if you could share the code of your animation library if you have written one. I was thinking that the 
API of ganim8 needs to be improved anyway.

**Comment by setanarut** [comment]
> I would greatly appreciate if you could share the code of your animation library if you have written one. I was thinking that the API of ganim8 needs to be improved anyway.

I published a working prototype today (on my alternative account). Pull requests and suggestions are welcome. I didn't use the % operator I mentioned before. It uses a simpler formula.

https://github.com/sokartes/ape

Demo:

```Bash
go run github.com/sokartes/ape/examples/runner@latest
```

**Comment by yohamta** [comment]
Thanks for sharing the code. It looks really nice and solid. I might create a PR if there's anything I noticed.

**Comment by flamendless** [comment]
The link https://github.com/sokartes/ape does not work though

**Comment by setanarut** [comment]
> The link https://github.com/sokartes/ape does not work though

I deleted my second account, a more updated version is available at this link. https://github.com/setanarut/bulimia/blob/main/engine%2Fanimplayer.go

for static sprites just set frameCount=1. example is in bulimia game.

**Comment by Palessan** [comment]
> Animation player should not be responsible for the screen drawing. It only needs to return `*ebiten.Image` with changing index.
> 
> for example:
> 
> ```go
> func (g *Game) Draw(screen *ebiten.Image) {
> 	screen.DrawImage(animation.GetFrame(), MyDrawImageOptions)
> }
> ```

Yes i came up with the same issue. I would prefer to be able to just get the next frame and draw it my way which already handles rotations etc..

**Comment by captkirk88** [comment]
> Animation player should not be responsible for the screen drawing. It only needs to return `*ebiten.Image` with changing index.
> 
> for example:
> 
> func (g *Game) Draw(screen *ebiten.Image) {
> 	screen.DrawImage(animation.GetFrame(), MyDrawImageOptions)
> }

This approach would also help with the new colorm package in ebitengine and result in #25 being a closed issue with no further action needed.  This is also preferable since ganim8 should not be drawing anything and only deal with the animation data.

**Comment by yottahmd** [comment]
Yes, that sounds good to me. Can we do this without breaking backward compatibility and mark the current API as deprecated?

**Comment by setanarut** [comment]
Here is my package. Animation playback state can be serialized with PlaybackData struct https://github.com/setanarut/anim

**Comment by activey** [comment]
In my case I use ganim8 not only for animations but also for static sprites rendering based on grid coordinates ;) It's just ... convenient. Please don't take it ;D

**Comment by activey** [comment]
> > Animation player should not be responsible for the screen drawing. It only needs to return `*ebiten.Image` with changing index.
> > for example:
> > func (g *Game) Draw(screen *ebiten.Image) {
> > screen.DrawImage(animation.GetFrame(), MyDrawImageOptions)
> > }
> 
> This approach would also help with the new colorm package in ebitengine and result in [#25](https://github.com/yottahmd/ganim8-lib/issues/25) being a closed issue with no further action needed. This is also preferable since ganim8 should not be drawing anything and only deal with the animation data.

Well, im afraid not entirely. New colorm.ColorM requires to use colorm.DrawImage in order to use it at all, so even when you just use ebiten.DrawImageOptions it simply wont work. Even more - colorm uses its own DrawImageOptions ;)

**Comment by setanarut** [comment]
> > > Animation player should not be responsible for the screen drawing. It only needs to return `*ebiten.Image` with changing index.
> > > for example:
> > > func (g *Game) Draw(screen *ebiten.Image) {
> > > screen.DrawImage(animation.GetFrame(), MyDrawImageOptions)
> > > }
> > 
> > This approach would also help with the new colorm package in ebitengine and result in [#25](https://github.com/yottahmd/ganim8-lib/issues/25) being a closed issue with no further action needed. This is also preferable since ganim8 should not be drawing anything and only deal with the animation data.
> 
> Well, im afraid not entirely. New colorm.ColorM requires to use colorm.DrawImage in order to use it at all, so even when you just use ebiten.DrawImageOptions it simply wont work. Even more - colorm uses its own DrawImageOptions ;) 
> 

My anim package is compatible with the colorm package. Also, please check out my Aseprite player package: https://github.com/setanarut/aseplayer

**Comment by yohamta0** [comment]
I'm closing this issue as it is addressed in the latest version.
